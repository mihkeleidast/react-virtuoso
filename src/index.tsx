export * from './components'
export * from './interfaces'
export { LogLevel } from './loggerSystem'
export * from './utils/context'
