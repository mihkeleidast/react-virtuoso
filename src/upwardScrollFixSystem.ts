import * as u from '@virtuoso.dev/urx'
import { domIOSystem } from './domIOSystem'
import { listStateSystem } from './listStateSystem'
import { sizeSystem } from './sizeSystem'
import { UP, stateFlagsSystem } from './stateFlagsSystem'
import { ListItem } from './interfaces'
import { loggerSystem, LogLevel } from './loggerSystem'
import { simpleMemoize } from './utils/simpleMemoize'
import { recalcSystem } from './recalcSystem'

const isMobileSafari = simpleMemoize(() => {
  return /iP(ad|hone|od).+Version\/[\d.]+.*Safari/i.test(navigator.userAgent)
})

type UpwardFixState = [number, ListItem<any>[], number, number]
/**
 * Fixes upward scrolling by calculating and compensation from changed item heights, using scrollBy.
 */
export const upwardScrollFixSystem = u.system(
  ([
    { scrollBy, scrollTop, deviation, scrollingInProgress },
    { isScrolling, isAtBottom, scrollDirection, lastJumpDueToItemResize },
    { listState },
    { beforeUnshiftWith, shiftWithOffset, sizes, gap },
    { log },
    { recalcInProgress },
  ]) => {
    const deviationOffset = u.streamFromEmitter(
      u.pipe(
        listState,
        u.withLatestFrom(lastJumpDueToItemResize),
        u.scan(
          ([, prevItems, prevTotalCount, prevTotalHeight], [{ items, totalCount, bottom, offsetBottom }, lastJumpDueToItemResize]) => {
            const totalHeight = bottom + offsetBottom

            let newDev = 0
            if (prevTotalCount === totalCount) {
              if (prevItems.length > 0 && items.length > 0) {
                const atStart = items[0].originalIndex === 0 && prevItems[0].originalIndex === 0
                if (!atStart) {
                  newDev = totalHeight - prevTotalHeight
                  if (newDev !== 0) {
                    newDev += lastJumpDueToItemResize
                  }
                }
              }
            }

            return [newDev, items, totalCount, totalHeight] as UpwardFixState
          },
          [0, [], 0, 0] as UpwardFixState
        ),
        u.filter(([amount]) => amount !== 0),
        u.withLatestFrom(scrollTop, scrollDirection, scrollingInProgress, isAtBottom, log),
        u.filter(([, scrollTop, scrollDirection, scrollingInProgress]) => {
          return !scrollingInProgress && scrollTop !== 0 && scrollDirection === UP
        }),
        u.map(([[amount], , , , , log]) => {
          log('Upward scrolling compensation', { amount }, LogLevel.DEBUG)
          return amount
        })
      )
    )

    function scrollByWith(offset: number) {
      if (offset > 0) {
        u.publish(scrollBy, { top: -offset, behavior: 'auto' })
        u.publish(deviation, 0)
      } else {
        u.publish(deviation, 0)
        u.publish(scrollBy, { top: -offset, behavior: 'auto' })
      }
    }

    u.subscribe(u.pipe(deviationOffset, u.withLatestFrom(deviation, isScrolling)), ([offset, deviationAmount, isScrolling]) => {
      if (isScrolling && isMobileSafari()) {
        u.publish(deviation, deviationAmount - offset)
      } else {
        scrollByWith(-offset)
      }
    })

    // this hack is only necessary for mobile safari which does not support scrollBy while scrolling is in progress.
    // when the browser stops scrolling, restore the position and reset the glitching
    u.subscribe(
      u.pipe(
        u.combineLatest(u.statefulStreamFromEmitter(isScrolling, false), deviation, recalcInProgress),
        u.filter(([is, deviation, recalc]) => !is && !recalc && deviation !== 0),
        u.map(([_, deviation]) => deviation),
        u.throttleTime(1)
      ),
      scrollByWith
    )

    u.connect(
      u.pipe(
        shiftWithOffset,
        u.map((offset) => {
          return { top: -offset }
        })
      ),
      scrollBy
    )

    u.subscribe(
      u.pipe(
        beforeUnshiftWith,
        u.withLatestFrom(sizes, gap),
        u.map(([offset, { lastSize }, gap]) => offset * lastSize + offset * gap)
      ),
      (offset) => {
        u.publish(deviation, offset)
        requestAnimationFrame(() => {
          u.publish(scrollBy, { top: offset })
          requestAnimationFrame(() => {
            u.publish(deviation, 0)
            u.publish(recalcInProgress, false)
          })
        })
      }
    )

    return { deviation }
  },
  u.tup(domIOSystem, stateFlagsSystem, listStateSystem, sizeSystem, loggerSystem, recalcSystem)
)
