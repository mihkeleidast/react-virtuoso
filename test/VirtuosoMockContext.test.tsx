import { render } from '@testing-library/react'
import * as React from 'react'
import { Virtuoso, VirtuosoMockContext, TableVirtuoso } from '../src'

describe('VirtuosoMockContext', () => {
  type Item = { id: string; value: string }
  const data: Item[] = [
    { id: '1', value: 'foo' },
    { id: '2', value: 'bar' },
    { id: '3', value: 'baz' },
    { id: '4', value: 'ban' },
  ]

  describe('List', () => {
    it('correctly renders items', () => {
      const { container } = render(<Virtuoso data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })

    it('correctly renders items with useWindowScroll', () => {
      const { container } = render(<Virtuoso data={data} useWindowScroll />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })
  })

  describe('Table', () => {
    it('correctly renders items', () => {
      const { container } = render(<TableVirtuoso data={data} />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })

    it('correctly renders items with useWindowScroll', () => {
      const { container } = render(<TableVirtuoso data={data} useWindowScroll />, {
        wrapper: ({ children }) => (
          <VirtuosoMockContext.Provider value={{ viewportHeight: 300, itemHeight: 100 }}>{children}</VirtuosoMockContext.Provider>
        ),
      })

      expect(container).toMatchSnapshot()
    })
  })
})
