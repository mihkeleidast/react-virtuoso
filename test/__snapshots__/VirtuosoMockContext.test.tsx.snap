// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtuosoMockContext Grid correctly renders items 1`] = `
<div>
  <div
    data-test-id="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        class="virtuoso-grid-list"
        style="padding-top: 0px; padding-bottom: 100px;"
      >
        <div
          class="virtuoso-grid-item"
          data-index="0"
        >
          Item 0
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="1"
        >
          Item 1
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="2"
        >
          Item 2
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="3"
        >
          Item 3
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="4"
        >
          Item 4
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="5"
        >
          Item 5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext Grid correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 300px;"
  >
    <div
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        class="virtuoso-grid-list"
        style="padding-top: 0px; padding-bottom: 100px;"
      >
        <div
          class="virtuoso-grid-item"
          data-index="0"
        >
          Item 0
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="1"
        >
          Item 1
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="2"
        >
          Item 2
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="3"
        >
          Item 3
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="4"
        >
          Item 4
        </div>
        <div
          class="virtuoso-grid-item"
          data-index="5"
        >
          Item 5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext List correctly renders items 1`] = `
<div>
  <div
    data-test-id="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      data-viewport-type="element"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        data-test-id="virtuoso-item-list"
        style="box-sizing: border-box; padding-top: 0px; padding-bottom: 100px; margin-top: 0px;"
      >
        <div
          data-index="0"
          data-item-index="0"
          data-known-size="100"
        >
          Item 0
        </div>
        <div
          data-index="1"
          data-item-index="1"
          data-known-size="100"
        >
          Item 1
        </div>
        <div
          data-index="2"
          data-item-index="2"
          data-known-size="100"
        >
          Item 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext List correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 400px;"
  >
    <div
      data-viewport-type="window"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <div
        data-test-id="virtuoso-item-list"
        style="box-sizing: border-box; padding-top: 0px; padding-bottom: 100px; margin-top: 0px;"
      >
        <div
          data-index="0"
          data-item-index="0"
          data-known-size="100"
        >
          Item 0
        </div>
        <div
          data-index="1"
          data-item-index="1"
          data-known-size="100"
        >
          Item 1
        </div>
        <div
          data-index="2"
          data-item-index="2"
          data-known-size="100"
        >
          Item 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext Table correctly renders items 1`] = `
<div>
  <div
    data-test-id="virtuoso-scroller"
    data-virtuoso-scroller="true"
    style="height: 100%; outline: none; overflow-y: auto; position: relative;"
    tabindex="0"
  >
    <div
      data-viewport-type="element"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <table
        style="border-spacing: 0;"
      >
        <tbody
          data-test-id="virtuoso-item-list"
        >
          <tr
            data-index="0"
            data-item-index="0"
            data-known-size="100"
          >
            <td>
              Item $
              0
            </td>
          </tr>
          <tr
            data-index="1"
            data-item-index="1"
            data-known-size="100"
          >
            <td>
              Item $
              1
            </td>
          </tr>
          <tr
            data-index="2"
            data-item-index="2"
            data-known-size="100"
          >
            <td>
              Item $
              2
            </td>
          </tr>
          <tr>
            <td
              style="height: 100px; padding: 0px; border: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`VirtuosoMockContext Table correctly renders items with useWindowScroll 1`] = `
<div>
  <div
    data-virtuoso-scroller="true"
    style="position: relative; height: 400px;"
  >
    <div
      data-viewport-type="window"
      style="width: 100%; height: 100%; position: absolute; top: 0px;"
    >
      <table
        style="border-spacing: 0;"
      >
        <tbody
          data-test-id="virtuoso-item-list"
        >
          <tr
            data-index="0"
            data-item-index="0"
            data-known-size="100"
          >
            <td>
              Item $
              0
            </td>
          </tr>
          <tr
            data-index="1"
            data-item-index="1"
            data-known-size="100"
          >
            <td>
              Item $
              1
            </td>
          </tr>
          <tr
            data-index="2"
            data-item-index="2"
            data-known-size="100"
          >
            <td>
              Item $
              2
            </td>
          </tr>
          <tr>
            <td
              style="height: 100px; padding: 0px; border: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
